src/EFCore.PG/Query/Expressions/Internal/ArrayAnyAllExpression.cs

```diff
@@ -0,0 +1,156 @@
+            ArrayComparisonType arrayComparisonType,
+            [NotNull] string operatorSymbol,
+            [NotNull] Expression operand,
+            [NotNull] Expression array)
+        {
+            Check.NotNull(operand, nameof(operand));
+            Check.NotNull(array, nameof(array));
+            Check.NotNull(array, nameof(operatorSymbol));
+
+            Operand = operand;
+            Operator = operatorSymbol;
+            Array = array;
+            ArrayComparisonType = arrayComparisonType;
+        }
+
+        /// <inheritdoc />
+        protected override Expression Accept(ExpressionVisitor visitor)
+            => visitor is NpgsqlQuerySqlGenerator npsgqlGenerator
+                ? npsgqlGenerator.VisitArrayAnyAll(this)
+                : base.Accept(visitor);
+
+        /// <inheritdoc />
+        protected override Expression VisitChildren(ExpressionVisitor visitor)
+        {
+            if (!(visitor.Visit(Operand) is Expression operand))
+                throw new ArgumentException($"The {nameof(operand)} of a {nameof(ArrayAnyAllExpression)} cannot be null.");
+
+            if (!(visitor.Visit(Array) is Expression collection))
+                throw new ArgumentException($"The {nameof(collection)} of a {nameof(ArrayAnyAllExpression)} cannot be null.");
+
+            return
+                operand == Operand && collection == Array
+                    ? this
+                    : new ArrayAnyAllExpression(ArrayComparisonType, Operator, operand, collection);
+        }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+            => obj is ArrayAnyAllExpression likeAnyExpression && Equals(likeAnyExpression);
+
+        /// <inheritdoc />
+        public bool Equals(ArrayAnyAllExpression other)
+            => Operand.Equals(other?.Operand) &&
+               Operator.Equals(other?.Operator) &&
+               ArrayComparisonType.Equals(other?.ArrayComparisonType) &&
+               Array.Equals(other?.Array);
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+            => unchecked((397 * Operand.GetHashCode()) ^
+                         (397 * Operator.GetHashCode()) ^
+                         (397 * ArrayComparisonType.GetHashCode()) ^
+                         (397 * Array.GetHashCode()));
+
+        /// <inheritdoc />
+        public override string ToString()
+            => $"{Operand} {Operator} {ArrayComparisonType.ToString()} ({Array})";
+    }
+
+    /// <summary>
+    /// Represents whether an array comparison is ANY or ALL.
+    /// </summary>
+    public enum ArrayComparisonType : byte
+    {
+        // ReSharper disable once InconsistentNaming
+        /// <summary>
+        /// Represents an ANY array comparison.
+        /// </summary>
+        ANY = 0,
+
+        // ReSharper disable once InconsistentNaming
+        /// <summary>
+        /// Represents an ALL array comparison.
+        /// </summary>
+        ALL = 1 << 0
```

Comment:
I don't think there's any reason to give values to this enum's members, as they're not meant to be persisted or transferred anywhere (and I'm curious as to the `<< 0` - you're really used to flag enums? :)).

Reply:
lol... bad habit of compulsively incorporating flag enums... but definitely irrelevant here. 

I've dropped the assignments. I think I was considering some uninitialized condition when I drafted it, but simplified to avoid it.